main.tf

```diff
@@ -14,100 +14,162 @@
-  project         = "${var.project}"
-  count           = "${length(var.backend_params)}"
-  name            = "${var.name}-backend-${count.index}"
-  port_name       = "${element(split(",", element(var.backend_params, count.index)), 1)}"
-  protocol        = "${var.backend_protocol}"
-  timeout_sec     = "${element(split(",", element(var.backend_params, count.index)), 3)}"
-  backend         = ["${var.backends["${count.index}"]}"]
-  health_checks   = ["${element(google_compute_http_health_check.default.*.self_link, count.index)}"]
-  security_policy = "${var.security_policy}"
-  enable_cdn      = "${var.cdn}"
+  for_each = toset(keys(var.backends))
+
+  project     = var.project
+  #  name        = "${var.name}-backend-${split(",", each.value)[1]}-${split(",", each.value)[2]}"
+  name        = "${var.name}-backend-${each.value}"
+
+  port_name   = local.bp_maps[tonumber(each.value)]["port_name"]
+  protocol    = var.backend_protocol
+  timeout_sec = local.bp_maps[tonumber(each.value)]["timeout"]
+
+  dynamic "backend" {
+    # keeping the old interface(var.backends) looks like overkill here
+    for_each = var.backends[each.value]
+    content {
+      balancing_mode               = lookup(backend.value, "balancing_mode", null)
+      capacity_scaler              = lookup(backend.value, "capacity_scaler", null)
+      description                  = lookup(backend.value, "description", null)
+      group                        = lookup(backend.value, "group", null)
+      max_connections              = lookup(backend.value, "max_connections", null)
+      max_connections_per_instance = lookup(backend.value, "max_connections_per_instance", null)
+      max_rate                     = lookup(backend.value, "max_rate", null)
+      max_rate_per_instance        = lookup(backend.value, "max_rate_per_instance", null)
+      max_utilization              = lookup(backend.value, "max_utilization", null)
+    }
+  }
+
+  health_checks = [
+    google_compute_health_check.default[tonumber(each.value)].self_link]
+
+  security_policy = var.security_policy
+  enable_cdn      = var.cdn
 }
 
-resource "google_compute_http_health_check" "default" {
-  project      = "${var.project}"
-  count        = "${length(var.backend_params)}"
-  name         = "${var.name}-backend-${count.index}"
-  request_path = "${element(split(",", element(var.backend_params, count.index)), 0)}"
-  port         = "${element(split(",", element(var.backend_params, count.index)), 2)}"
+resource "google_compute_health_check" "default" {
+  # Note:
+  #  - tf supports only sets of stings, so options below don't work
+  #     for_each = local.bp_maps # - not supported type
+  #     for_each = toset(range(length(local.bp_maps))) # - not supported type
+  #     for_each = toset(keys(var.backends)) # requires calling type conversion function
+  #  - keeping the old interface(var.backed_params) is a bit too expensive
+  # without changing the itnerface count.index works better than for_each
```

Comment:
Can (should?) we change the interface and embed the health check into the backends variable instead? This would then remove `bq_maps` completely and you can just do `for_each = var.backends` +  `lookup`


```terraform
backends = { 
  "foo" = {
    groups = [
      {
        group = ""
      },
      {
        group = ""
      },      
    ]

    health_check = {
      path = "/"
      port = 8080
      host = null
    }
  }

}
```

Reply:
@Dev25, I agree and I suggested the same. 
However the whole `for_each` changes are set aside in a `for_each` branch as @morgante advised. So we're trying to finalize everything but that, as the `for_each` [bug](https://github.com/hashicorp/terraform/issues/22735) brakes 2 of multi backends examples

Reply:
Right fair enough, that seems reasonable.